Tomcat 7 and Rundeck v3 compatibility

Retrieving execution log using rundeck-s3-log-plugin after upgrading to Rundeck 3.0.7 fails with the following error in the catalina log:
`org.springframework.web.util.NestedServletException: Handler dispatch failed; nested exception is java.lang.NoClassDefFoundError: javax/servlet/WriteListener`

This works previously on Rundeck 2.11.4.  Looking around it seems as though there are issues with upgraded grails version in Rundeck v3 and tomcat 7.

https://stackoverflow.com/questions/46106614/grails-3-3-0-on-tomcat-7-0-57
https://stackoverflow.com/questions/49990692/grails-application-response-outputstream-fails-since-upgrade-to-grails-3-3


* Rundeck version: 3.0.7
* install type: war
* OS Name/version: RHEL 6.10
* DB Type/version: mysql
* Tomcat: 7.0.69.0

**To Reproduce**
View log output of completed execution and select Text/HTML/Download option to retrieve log from S3 log storage. Get following error:

> Handler dispatch failed; nested exception is java.lang.NoClassDefFoundError: javax/servlet/WriteListener


```
Full Stack:

2019-02-13 14:44:13.293 ERROR --- [io-8080-exec-20] StackTrace                               : Full Stack Trace:

org.springframework.web.util.NestedServletException: Handler dispatch failed; nested exception is java.lang.NoClassDefFoundError: javax/servlet/WriteListener
>...at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:982)
>...at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901)
>...at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
>...at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:861)
>...at javax.servlet.http.HttpServlet.service(HttpServlet.java:624)
>...at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
>...at javax.servlet.http.HttpServlet.service(HttpServlet.java:731)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.apache.catalina.filters.RequestDumperFilter.doFilter(RequestDumperFilter.java:203)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)                                                                                                                                                                                                                            
>...at org.apache.catalina.filters.ExpiresFilter.doFilter(ExpiresFilter.java:1201)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at eu.fidessa.web.NoCacheFilter.doFilter(NoCacheFilter.java:51)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.springframework.security.web.jaasapi.JaasApiIntegrationFilter$1.run(JaasApiIntegrationFilter.java:96)
>...at java.security.AccessController.doPrivileged(Native Method)
>...at javax.security.auth.Subject.doAs(Subject.java:422)
>...at org.springframework.security.web.jaasapi.JaasApiIntegrationFilter.doFilter(JaasApiIntegrationFilter.java:105)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.springframework.boot.web.filter.ApplicationContextHeaderFilter.doFilterInternal(ApplicationContextHeaderFilter.java:55)
>...at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.springframework.boot.actuate.trace.WebRequestTraceFilter.doFilterInternal(WebRequestTraceFilter.java:111)
>...at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
>...at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
>...at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:114)
>...at grails.plugin.springsecurity.web.UpdateRequestContextHolderExceptionTranslationFilter.doFilter(UpdateRequestContextHolderExceptionTranslationFilter.groovy:64)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at grails.plugin.springsecurity.web.filter.GrailsAnonymousAuthenticationFilter.doFilter(GrailsAnonymousAuthenticationFilter.groovy:54)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter(RememberMeAuthenticationFilter.java:158)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at org.springframework.security.web.jaasapi.JaasApiIntegrationFilter$1.run(JaasApiIntegrationFilter.java:96)
>...at java.security.AccessController.doPrivileged(Native Method)
>...at javax.security.auth.Subject.doAs(Subject.java:422)
>...at org.springframework.security.web.jaasapi.JaasApiIntegrationFilter.doFilter(JaasApiIntegrationFilter.java:105)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at grails.plugin.springsecurity.web.authentication.logout.MutableLogoutFilter.doFilter(MutableLogoutFilter.groovy:64)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at grails.plugin.springsecurity.web.SecurityRequestHolderFilter.doFilter(SecurityRequestHolderFilter.groovy:58)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
>...at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.grails.web.servlet.mvc.GrailsWebRequestFilter.doFilterInternal(GrailsWebRequestFilter.java:77)
>...at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.grails.web.filters.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:67)
>...at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:197)
>...at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.springframework.web.filter.CorsFilter.doFilterInternal(CorsFilter.java:96)
>...at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.springframework.boot.actuate.autoconfigure.MetricsFilter.doFilterInternal(MetricsFilter.java:103)
>...at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.springframework.boot.web.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:130)
>...at org.springframework.boot.web.support.ErrorPageFilter.access$000(ErrorPageFilter.java:66)
>...at org.springframework.boot.web.support.ErrorPageFilter$1.doFilterInternal(ErrorPageFilter.java:105)
>...at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
>...at org.springframework.boot.web.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:123)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)
>...at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)
>...at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:505)
>...at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:169)
>...at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)
>...at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:956)
>...at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
>...at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:436)
>...at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1078)
>...at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:625)
>...at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1749)
>...at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1708)
>...at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
>...at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
>...at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
>...at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NoClassDefFoundError: javax/servlet/WriteListener
>...at java.lang.Class.getDeclaredMethods0(Native Method)
>...at java.lang.Class.privateGetDeclaredMethods(Class.java:2701)
>...at java.lang.Class.privateGetPublicMethods(Class.java:2902)
>...at java.lang.Class.getMethods(Class.java:1615)
>...at java.beans.Introspector.getPublicDeclaredMethods(Introspector.java:1336)
>...at java.beans.Introspector.getTargetMethodInfo(Introspector.java:1197)
>...at java.beans.Introspector.getBeanInfo(Introspector.java:426)
>...at java.beans.Introspector.getBeanInfo(Introspector.java:173)
>...at groovy.lang.MetaClassImpl$15.run(MetaClassImpl.java:3328)
>...at java.security.AccessController.doPrivileged(Native Method)
>...at groovy.lang.MetaClassImpl.addProperties(MetaClassImpl.java:3326)
>...at groovy.lang.MetaClassImpl.initialize(MetaClassImpl.java:3303)
>...at org.codehaus.groovy.reflection.ClassInfo.getMetaClassUnderLock(ClassInfo.java:289)
>...at org.codehaus.groovy.reflection.ClassInfo.getMetaClass(ClassInfo.java:331)
>...at org.codehaus.groovy.reflection.ClassInfo.getMetaClass(ClassInfo.java:341)
>...at org.codehaus.groovy.runtime.metaclass.MetaClassRegistryImpl.getMetaClass(MetaClassRegistryImpl.java:281)
>...at org.codehaus.groovy.runtime.InvokerHelper.getMetaClass(InvokerHelper.java:901)
>...at org.codehaus.groovy.runtime.callsite.CallSiteArray.createPojoSite(CallSiteArray.java:124)
>...at org.codehaus.groovy.runtime.callsite.CallSiteArray.createCallSite(CallSiteArray.java:165)
>...at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
>...at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
>...at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:128)
>...at rundeck.controllers.ExecutionController$_downloadOutput_closure28.doCall(ExecutionController.groovy:795)
>...at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
>...at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
>...at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>...at java.lang.reflect.Method.invoke(Method.java:498)
>...at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:98)
>...at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)                                                                                                                                                                                                                                                           
>...at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:264)
>...at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1034)
>...at groovy.lang.Closure.call(Closure.java:418)
>...at groovy.lang.Closure.call(Closure.java:434)
>...at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2125)
>...at org.codehaus.groovy.runtime.dgm$162.invoke(Unknown Source)
>...at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoMetaMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:251)
>...at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.call(PogoMetaMethodSite.java:71)
>...at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
>...at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
>...at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:128)
>...at rundeck.controllers.ExecutionController.downloadOutput(ExecutionController.groovy:784)
>...at org.grails.core.DefaultGrailsControllerClass$MethodHandleInvoker.invoke(DefaultGrailsControllerClass.java:223)
>...at org.grails.core.DefaultGrailsControllerClass.invoke(DefaultGrailsControllerClass.java:188)
>...at org.grails.web.mapping.mvc.UrlMappingsInfoHandlerAdapter.handle(UrlMappingsInfoHandlerAdapter.groovy:90)
>...at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:967)
>...... 107 common frames omitted
Caused by: java.lang.ClassNotFoundException: javax.servlet.WriteListener
>...at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1858)
>...at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1701)
>...... 152 common frames omitted

```

@sjrd218  I've updated original post. Full stack:

```
2019-02-13 14:44:13.293 ERROR --- [io-8080-exec-20] StackTrace                               : Full Stack Trace:

org.springframework.web.util.NestedServletException: Handler dispatch failed; nested exception is java.lang.NoClassDefFoundError: javax/servlet/WriteListener
>...at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:982)
>...at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901)
>...at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
>...at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:861)
>...at javax.servlet.http.HttpServlet.service(HttpServlet.java:624)
>...at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
>...at javax.servlet.http.HttpServlet.service(HttpServlet.java:731)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.apache.catalina.filters.RequestDumperFilter.doFilter(RequestDumperFilter.java:203)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)                                                                                                                                                                                                                            
>...at org.apache.catalina.filters.ExpiresFilter.doFilter(ExpiresFilter.java:1201)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at eu.fidessa.web.NoCacheFilter.doFilter(NoCacheFilter.java:51)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.springframework.security.web.jaasapi.JaasApiIntegrationFilter$1.run(JaasApiIntegrationFilter.java:96)
>...at java.security.AccessController.doPrivileged(Native Method)
>...at javax.security.auth.Subject.doAs(Subject.java:422)
>...at org.springframework.security.web.jaasapi.JaasApiIntegrationFilter.doFilter(JaasApiIntegrationFilter.java:105)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.springframework.boot.web.filter.ApplicationContextHeaderFilter.doFilterInternal(ApplicationContextHeaderFilter.java:55)
>...at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.springframework.boot.actuate.trace.WebRequestTraceFilter.doFilterInternal(WebRequestTraceFilter.java:111)
>...at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
>...at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
>...at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:114)
>...at grails.plugin.springsecurity.web.UpdateRequestContextHolderExceptionTranslationFilter.doFilter(UpdateRequestContextHolderExceptionTranslationFilter.groovy:64)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at grails.plugin.springsecurity.web.filter.GrailsAnonymousAuthenticationFilter.doFilter(GrailsAnonymousAuthenticationFilter.groovy:54)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter(RememberMeAuthenticationFilter.java:158)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at org.springframework.security.web.jaasapi.JaasApiIntegrationFilter$1.run(JaasApiIntegrationFilter.java:96)
>...at java.security.AccessController.doPrivileged(Native Method)
>...at javax.security.auth.Subject.doAs(Subject.java:422)
>...at org.springframework.security.web.jaasapi.JaasApiIntegrationFilter.doFilter(JaasApiIntegrationFilter.java:105)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at grails.plugin.springsecurity.web.authentication.logout.MutableLogoutFilter.doFilter(MutableLogoutFilter.groovy:64)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at grails.plugin.springsecurity.web.SecurityRequestHolderFilter.doFilter(SecurityRequestHolderFilter.groovy:58)
>...at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
>...at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
>...at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.grails.web.servlet.mvc.GrailsWebRequestFilter.doFilterInternal(GrailsWebRequestFilter.java:77)
>...at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.grails.web.filters.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:67)
>...at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:197)
>...at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.springframework.web.filter.CorsFilter.doFilterInternal(CorsFilter.java:96)
>...at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.springframework.boot.actuate.autoconfigure.MetricsFilter.doFilterInternal(MetricsFilter.java:103)
>...at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.springframework.boot.web.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:130)
>...at org.springframework.boot.web.support.ErrorPageFilter.access$000(ErrorPageFilter.java:66)
>...at org.springframework.boot.web.support.ErrorPageFilter$1.doFilterInternal(ErrorPageFilter.java:105)
>...at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
>...at org.springframework.boot.web.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:123)
>...at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
>...at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
>...at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)
>...at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)
>...at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:505)
>...at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:169)
>...at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)
>...at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:956)
>...at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
>...at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:436)
>...at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1078)
>...at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:625)
>...at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1749)
>...at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1708)
>...at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
>...at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
>...at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
>...at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NoClassDefFoundError: javax/servlet/WriteListener
>...at java.lang.Class.getDeclaredMethods0(Native Method)
>...at java.lang.Class.privateGetDeclaredMethods(Class.java:2701)
>...at java.lang.Class.privateGetPublicMethods(Class.java:2902)
>...at java.lang.Class.getMethods(Class.java:1615)
>...at java.beans.Introspector.getPublicDeclaredMethods(Introspector.java:1336)
>...at java.beans.Introspector.getTargetMethodInfo(Introspector.java:1197)
>...at java.beans.Introspector.getBeanInfo(Introspector.java:426)
>...at java.beans.Introspector.getBeanInfo(Introspector.java:173)
>...at groovy.lang.MetaClassImpl$15.run(MetaClassImpl.java:3328)
>...at java.security.AccessController.doPrivileged(Native Method)
>...at groovy.lang.MetaClassImpl.addProperties(MetaClassImpl.java:3326)
>...at groovy.lang.MetaClassImpl.initialize(MetaClassImpl.java:3303)
>...at org.codehaus.groovy.reflection.ClassInfo.getMetaClassUnderLock(ClassInfo.java:289)
>...at org.codehaus.groovy.reflection.ClassInfo.getMetaClass(ClassInfo.java:331)
>...at org.codehaus.groovy.reflection.ClassInfo.getMetaClass(ClassInfo.java:341)
>...at org.codehaus.groovy.runtime.metaclass.MetaClassRegistryImpl.getMetaClass(MetaClassRegistryImpl.java:281)
>...at org.codehaus.groovy.runtime.InvokerHelper.getMetaClass(InvokerHelper.java:901)
>...at org.codehaus.groovy.runtime.callsite.CallSiteArray.createPojoSite(CallSiteArray.java:124)
>...at org.codehaus.groovy.runtime.callsite.CallSiteArray.createCallSite(CallSiteArray.java:165)
>...at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
>...at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
>...at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:128)
>...at rundeck.controllers.ExecutionController$_downloadOutput_closure28.doCall(ExecutionController.groovy:795)
>...at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
>...at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
>...at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>...at java.lang.reflect.Method.invoke(Method.java:498)
>...at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:98)
>...at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)                                                                                                                                                                                                                                                           
>...at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:264)
>...at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1034)
>...at groovy.lang.Closure.call(Closure.java:418)
>...at groovy.lang.Closure.call(Closure.java:434)
>...at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2125)
>...at org.codehaus.groovy.runtime.dgm$162.invoke(Unknown Source)
>...at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoMetaMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:251)
>...at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.call(PogoMetaMethodSite.java:71)
>...at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
>...at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
>...at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:128)
>...at rundeck.controllers.ExecutionController.downloadOutput(ExecutionController.groovy:784)
>...at org.grails.core.DefaultGrailsControllerClass$MethodHandleInvoker.invoke(DefaultGrailsControllerClass.java:223)
>...at org.grails.core.DefaultGrailsControllerClass.invoke(DefaultGrailsControllerClass.java:188)
>...at org.grails.web.mapping.mvc.UrlMappingsInfoHandlerAdapter.handle(UrlMappingsInfoHandlerAdapter.groovy:90)
>...at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:967)
>...... 107 common frames omitted
Caused by: java.lang.ClassNotFoundException: javax.servlet.WriteListener
>...at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1858)
>...at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1701)
>...... 152 common frames omitted
```

Thanks

--- a/rundeckapp/grails-app/controllers/rundeck/controllers/ExecutionController.groovy
+++ b/rundeckapp/grails-app/controllers/rundeck/controllers/ExecutionController.groovy
@@ -40,6 +40,7 @@ import com.dtolabs.rundeck.plugins.logs.ContentConverterPlugin
 import com.dtolabs.rundeck.server.authorization.AuthConstants
 import com.dtolabs.rundeck.util.MetricsStatsBuilder
 import grails.converters.JSON
+import groovy.transform.CompileStatic
 import org.quartz.JobExecutionContext
 import org.springframework.dao.DataAccessResourceFailureException
 import rundeck.CommandExec
@@ -748,7 +749,7 @@ class ExecutionController extends ControllerBase{
             response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR)
             def msg= g.message(code: reader.errorCode, args: reader.errorData)
             log.error("Output file reader error: ${msg}")
-            response.outputStream << msg
+            appendOutput(msg)
             return
         }else if (reader.state != ExecutionLogState.AVAILABLE) {
             //TODO: handle other states
@@ -783,8 +784,8 @@ class ExecutionController extends ControllerBase{
                 } catch (Exception exc) {
                 }
             }
-            response.outputStream << (isFormatted?"${logFormater.format(msgbuf.datetime)} [${msgbuf.metadata?.user}@${msgbuf.metadata?.node} ${msgbuf.metadata?.stepctx?:'_'}][${msgbuf.loglevel}] ${message}" : message)
-            response.outputStream<<lineSep
+            appendOutput((isFormatted?"${logFormater.format(msgbuf.datetime)} [${msgbuf.metadata?.user}@${msgbuf.metadata?.node} ${msgbuf.metadata?.stepctx?:'_'}][${msgbuf.loglevel}] ${message}" : message))
+            appendOutput(lineSep)
         }
         iterator.close()
     }
@@ -808,11 +809,12 @@ class ExecutionController extends ControllerBase{
             response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR)
             def msg= g.message(code: reader.errorCode, args: reader.errorData)
             log.error("Output file reader error: ${msg}")
-            response.outputStream << msg
+            appendOutput(msg)
             return
         }else if(reader.state == ExecutionLogState.WAITING){
             if(params.reload=='true') {
-                response.outputStream << '''<html>
+                response.setContentType("text/html")
+                appendOutput('''<html>
                 <head>
                 <title></title>
                 </head>
@@ -820,9 +822,9 @@ class ExecutionController extends ControllerBase{
                 <div class="container">
                 <div class="row">
                 <div class="col-sm-12">
-                '''
-                response.outputStream << g.message(code: "execution.html.waiting")
-                response.outputStream << '''
+                ''')
+                appendOutput(g.message(code: "execution.html.waiting"))
+                appendOutput('''
                 </div>
                 <script>
                 setTimeout(function(){
@@ -831,7 +833,7 @@ class ExecutionController extends ControllerBase{
                 </script>
                 </body>
                 </html>
-                '''
+                ''')
             }else{
                 response.setStatus(HttpServletResponse.SC_NOT_FOUND)
                 log.error("Output file not available")
@@ -852,7 +854,8 @@ class ExecutionController extends ControllerBase{
         def iterator = reader.reader
         iterator.openStream(0)
         def lineSep=System.getProperty("line.separator")
-        response.outputStream<<"""<html>
+        response.setContentType("text/html")
+        appendOutput("""<html>
 <head>
 <title></title>
 <link rel="stylesheet" href="${g.assetPath(src:'app.less.css')}"  />
@@ -862,7 +865,7 @@ class ExecutionController extends ControllerBase{
 <div class="container">
 <div class="row">
 <div class="col-sm-12">
-<div class="ansicolor ansicolor-${(params.ansicolor in ['false','off'])?'off':'on'}" >"""
+<div class="ansicolor ansicolor-${(params.ansicolor in ['false','off'])?'off':'on'}" >""")
 
         def csslevel=!(params.loglevels in ['off','false'])
         def renderContent = shouldConvertContent(params)
@@ -894,23 +897,23 @@ class ExecutionController extends ControllerBase{
             }
             def css="log_line" + (csslevel?" level_${msgbuf.loglevel.toString().toLowerCase()}":'')
 
-            response.outputStream << "<div class=\"$css\" >"
-            response.outputStream << msghtml
-            response.outputStream << '</div>'
+            appendOutput("<div class=\"$css\" >")
+            appendOutput(msghtml)
+            appendOutput('</div>')
 
-            response.outputStream<<lineSep
+            appendOutput(lineSep)
         }
         iterator.close()
         if(jobcomplete || params.reload!='true'){
-            response.outputStream << '''</div>
+            appendOutput('''</div>
 </div>
 </div>
 </div>
 </body>
 </html>
-'''
+''')
         }else{
-            response.outputStream << '''</div>
+            appendOutput('''</div>
 </div>
 </div>
 </div>
@@ -921,9 +924,14 @@ setTimeout(function(){
 </script>
 </body>
 </html>
-'''
+''')
         }
 
+    }
+
+    @CompileStatic
+    private void appendOutput(def output) {
+        response.outputStream << output
     }
 
     /**

